### PR TITLE
chore(benchmarks): add benchmark application in go

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR introduces a benchmark client application in Go. It uses the local Go client specifically, and currently only implements the worker command, which is similar to the worker benchmark application in Java. The idea here is to allow us to run benchmarks with both client types at some point. This came about from me needing to test a Go PR with the client against a broker. We should eventually add a starter, but this is a good starting point, I think.

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Review by miguelpires — CHANGES_REQUESTED** [review]
Just a few additional comments:
* Maybe we can improve the structure in `/projects` because, currently the Go code, the `go.mod` and `go.sum` files, the Docker files and the Java code are all next to each other. 
* Maybe it's just me, but it's not very clear how to use this. What do I do after building `zbench`? Maybe we could add a short explanation to the README stating what needs to be changed in the docker files to use `zbench`.
* Please `gofmt` the code.

**Comment by npepinpe** [comment]
So I opened this (arguably too early?) because I wrote a small application to test a user's code, and figured it'd be a shame to delete it as it could be useful again some day to have a test application written in Go.

The app is meant to be used as `zbench worker` or `zbench starter` and it would do like our `starter.jar` or `worker.jar`. I guess if we just used `zbench run` or `zbench allinone` or something we could simulate the `all-in-one.jar` as well?

But yeah you're probably right, this can use some polishing. Not particularly urgent though

**Comment by MiguelPires** [comment]
I agree that it's not a priority and I understand where this comes from but I think that at least documenting this would make it even more useful for others. I'd be more than glad to split the remaining work, if you want. I can make the code and structure changes and you can write update the README.

**Comment by npepinpe** [comment]
Yeah totally! Happy to get help but I also want to help! Generally I guess we'd need to expand this to include a starter equivalent, and make it compatible with the cloud (though I guess you can kind of do that already with env var so... :man_shrugging:).

We can have a quick Zoom call at some point this week to hash it out

**Comment by MiguelPires** [comment]
Hey @npepinpe, I made some overall improvements in terms of code and layout. Let me know what you think. We can merge this as a starting point and then build on it. Do you want to add a short explanation as to how to use zbench?

**Comment by npepinpe** [comment]
Yeah I'll do that, thanks :slightly_smiling_face:

**Comment by MiguelPires** [comment]
Ah I forgot to ask you this: what do you think about moving this into the `clients/go` folder, so we don't have to re-download all the dependencies in the repo?

**Comment by npepinpe** [comment]
I think we can also use the "replace" mechanism of the go module with a local path, no? I guess that's a little dirtier, but at the same time I'm not sure how I feel about placing a benchmark application under `clients/go` as if it were official :thinking:

**Comment by MiguelPires** [comment]
`go mod vendor` vendors dependencies regardless of whether they're local or not so `replace` doesn't prevent that. But I was actually talking about the go client's dependencies, they're duplicated again here (except if they're only in "internal" packages). But it's not a big deal if you want to keep it here, just thought I was worth bringing up

**Comment by npepinpe** [comment]
No, that's a good point, I hadn't thought of that. I'm just wondering how people will perceive it if it's under `clients/go`? I suppose if it's well documented enough it should be OK? wdyt?

File: benchmarks/project/cmd/zbench/internal/commands/root.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+		if !audienceExists {
+			if err := os.Setenv(zbc.OAuthTokenAudienceEnvVar, host); err != nil {
+				return err
+			}
+		}
+
+		providerConfig := zbc.OAuthProviderConfig{}
+
+		// create a credentials provider with the specified parameters
+		credsProvider, err = zbc.NewOAuthCredentialsProvider(&providerConfig)
```
Comment (MiguelPires) [inline benchmarks/project/cmd/zbench/internal/commands/root.go]:
```suggestion
		// create a credentials provider with the specified parameters
		credsProvider, err = zbc.NewOAuthCredentialsProvider(&zbc.OAuthProviderConfig{})
```

File: benchmarks/project/zbench/internal/commands/root.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+		if !audienceExists {
+			if err := os.Setenv(zbc.OAuthTokenAudienceEnvVar, host); err != nil {
+				return err
+			}
+		}
+
+		providerConfig := zbc.OAuthProviderConfig{}
+
+		// create a credentials provider with the specified parameters
+		credsProvider, err = zbc.NewOAuthCredentialsProvider(&providerConfig)
```
Comment (MiguelPires) [inline benchmarks/project/zbench/internal/commands/root.go:129]:
```suggestion
		// create a credentials provider with the specified parameters
		credsProvider, err = zbc.NewOAuthCredentialsProvider(&zbc.OAuthProviderConfig{})
```

**Comment by MiguelPires** [comment]
I think your reservation is that putting it in `clients/go` will make users draw comparisons with `zbctl`, except that's better tested and we're serious about maintaining it, right? I think that even if we put it next to `zbctl` and people consider it an "official" tool and ask for features or support, then great, it must be because it's generating value for them. We can then decide where those requests fall within our priorities at that moment but, at least it was useful for someone and they're trying to increase that value. So I don't think user perception is a problem here but I'm also fine with leaving it where it is.

**Comment by MiguelPires** [comment]
@npepinpe I removed the vendor folder and formatted the code again. Let me know if this is fine with you as is, so I can merge it

**Review by miguelpires — APPROVED** [review]
bors r+

**Pull request merged**